### ISSUE TITLE
Really outperform baselines that much ?

### ISSUE THREAD

**Issue by HolmesShuan** [kevinlin311tw/cvpr16-deepbit#17]
Hi, I have noticed that reported performance has far beyond baselines but which is `mAP@1K`. Consider your baselines are all `mAP` which should be lower than `mAP@1K` in your [implementation](https://github.com/kevinlin311tw/cvpr16-deepbit/blob/master/analysis/precision.m). As claimed in section 4.3, 
> Following the settings in [24](Deep hashing for compact binary codes learning, CVPR 15) ...

however, DH reported `mAP`  rather than `mAP@1K` (if you double check the setting in [24]). Could you please give us more details about the real `mAP` on Cifar10 to fully compare with others.    
In our evaluation, even ITQ (well-known and still competitive baseline) has reached **20+** `mAP@1K`, which makes us wonder whether Deepbit outperforms baselines that much?    

Anyway, good work and thanks for the open source codes.

**Comment by kevinlin311tw** [comment]
After double checking, I confirmed this problem. I am sorry that the eval criterions are not consistent. I will fix it and report new mAP here. 
Also, please do not compare the numbers in this CIFAR10 experiment since there has problem.

**Comment by kevinlin311tw** [comment]
Thank you for your interest in this work, and thanks for pointing out this issue. I have to update and discuss more about the mAP on CIFAR10 here.

First, according to my re-evaluation, I found that DeepBit’s mAP is not stable using their evaluation setting. To be specific, sometimes DeepBit outperforms DH a lot, but sometimes the performance is bad. After a few experiments, I conjecture that, the performance depends mainly on the choice of the randomly selected 1000 queries.

Instead of selecting only 1000 queries which may introduce the randomness to the evaluation, I decide to report the mAP using the entire test set (in a total of 10,000 queries). Note that I also used this strategy in my previous implementation, and more importantly it is reproducible for future development. Particularly, I just downloaded the models from this repo, and conducted the evaluation. By evaluating DeepBit’s performance with this stricter criterion, the mAP of DeepBit on CIFAR10 is not always good as shown below.

|     Method     |   16 bits   |   32 bits   |   64 bits   
|----------------|:-----------:|:-----------:|:-----------:
|      DeepBit (this repo)      |    14.34    |    16.32    |    17.96    
|      DeepBit (bugs fixed)     |    21.70    |    20.64    |    23.07

Why DeepBit performs worse than DH on CIFAR10? A possible reason is that the evaluation criterions are still inconsistent. Another key reason is mainly because the imperfect implementation in this repo. In fact, I noticed there is a few implementation bugs in this repo. First, the activation function of the latent layer does not match the proposed objective function. Second, I didn’t fully empower the third term of the proposed objective function. I just restricted it to rotation invariant descriptors.

After fixing the implementation by (1) replacing relu with sigmoid for the latent layer and (2) training with data augmentation (random combinations of rotation, scaling, and transition using [caffe-augmentation](https://github.com/kevinlin311tw/caffe-augmentation)), DeepBit learns more general and better descriptors. In addition, I expect the performance could be further boosted by doing hyperparameter selection for each individual term in the objective function.

Hope this may help!

**Comment by HolmesShuan** [comment]
It is interesting that `32bit` is worse than `16bit` in bug fixed version. Anyway, thanks for your help.  👍
